### ISSUE TITLE
Strict mode static property is undefined (no compiler error)

### ISSUE THREAD

**Issue by Doccos** [microsoft/TypeScript#40649]
**TypeScript Version:**  4.0.3
**IDE:**  PHPStorm 2020.2

**Search Terms:** strictPropertyInitialization static property undefined

**tsconfig.json**
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "sourceMap": true,
    "strict": true
  },
  "exclude": [
    "node_modules"
  ]
}
```
**Code**

```ts
class StaticUndefined {

    
    private static props: { 
        test: string,
        test2: number
    }; //should be undefined
    
    public static getProps() {
        console.log(typeof StaticUndefined.props); // is undefined // no error during compile or in IDE (strictPropertyInitialization not triggert)
        console.log(typeof StaticUndefined.props.test); //error
        console.log(StaticUndefined.props.test); //error
        
    }
}

StaticUndefined.getProps();
```

**Expected behavior:**
The compiler and IDE should warn me that "props" may be "undefined" and should check their status

**Actual behavior:**
It will compile and the IDE will not show an error, but the script will crash (StaticUndefined.props is undefined).

[Playground link](https://www.typescriptlang.org/play?ts=4.1.0-dev.20200919#code/MYGwhgzhAEDKAuZ4EtgFUB2ATApgM2Qxy2gG8Aoc6a6KmgBwCdkA3JHaCRFYaJge3oQAXGVo0J0eDi6iuzDAHMANHUlSZ8AEyiMAVwC2AIxyM10AL4BuaAHpbEABb89IEieh7s+QsXPn6PSMQVE5uUMUceAAFRkEIAAoASjJzCWB+DAh+EBwAOhB+RQT4AE96HH48OHD0bwIiLDyBISSbe2hkGC9cBuI7W2gMfmhTOMZoLD0FRWgMg3pkXOh+CcJoAEkAEQBRaAT5VBi4isYyjYxkFDAQgC8kZEyh-ngpZkVIs6S0mgysnPyhWKZQqVRqDzqvV8TRaEDy0i4bQGY1WP2of2yuQKRQSCAhmChjWaJzhCPgSPsKLM6mo5gs5Hp5DxPAJPiJkWO8WSVnIQA)

**Related Issues:**

**Comment by yw662** [comment]
This is a interesting one for me, and I tried something similar to this issue,

```ts
    private static props: { 
        test: string,
        test2: number
    } = undefined
```
or
```ts
    private static props: { 
        test: string,
        test2: number
    } = {}
```
Will raise the expected error.

For me this is like
```ts
let a: number
```
which is actually legal since you might assign to it later.

However,
```ts
    private static readonly props: { 
        test: string,
        test2: number
    }
``` 
won't raise the error either, when
```ts
const a: number
```
is not good.

And for instance members, like,
```ts
    private props: { 
        test: string,
        test2: number
    };
```
It always works as expected if they are not assigned in some constructors.

I guess there are some difficulties to let static mutable members be checked this way: If non-assigning declaration is disallowed for static mutable members, then non-assigning variable declaration should also be disallowed.
Besides, these mutable members, can be write-only members or setters. If a member is write-only, they should not be assigned on declaration. Although a `private writeonly foo` is not a `private foo`, which implies read-write, anyway.

However, static read-only members should be forced to be assigned, which is anyway a bug, and `const foo` is also forced to be assigned.

For the type check on the `getProps()` side, this is somewhat a working-as-intended: the variable is not declared optional, so they "cannot" be undefined.
For me there is a work around if you want the type check enforced on runtime: declare them as optional, like
```ts
    private static props?: { 
        test: string,
        test2: number
    };
```

They `getProps` will know they might be undefined.

**Comment by Doccos** [comment]
Yes, the workaround works, but is not perfect.
This should be done automatically to avoid mistakes.
So for
```ts
private static f: number;
```
It is declared as a number, but not initialized, so it should be of type "number | undefined" but is of type "number".

**Comment by yw662** [comment]
For me I would rather non-assigning non-optional declaration being disallowed. That's to say, the question mark in `static foo?`, and `undefined` in `let a: number | undefined` should be enforced.

However this is too much as a breaking change.

**Comment by jack-williams** [comment]
Duplicate of #27899?

**Comment by RyanCavanaugh** [comment]
👆
